### PR TITLE
Allowing use of memoization array.

### PR DESCRIPTION
**Pull request by nmn**
As explained in #30 

Hooks usually accept a second argument of values to memoize over. In that respect, the `useMappedState` is a little weird in that it accepts the mapper function to be pre-memoized or defined as a constant outside of the component.

Accepting an argument of values to memoize based on removes the need to do that.

By defaulting to reference to the mapper function, we get existing semantics when no second argument is provided and the change is backwards compatible.

### CODE REVIEW

**Comment by nmn** [comment]
This commit doesn't go far enough. Some of the logic for recalculating the state would need to be changed as well. Opening the PR more as an example and not necessarily to be merged.

File: src/index.ts [hunk 1]
```diff
@@ -87,7 +78,7 @@ export function useMappedState<TState, TResult>(
         unsubscribe();
       };
     },
-    [store, mapState],
+    [store, ...memoizationArray],
```
Comment (ianobermiller) [inline src/index.ts]: I don't think we want this here, as it would cause the store to re-render every time one of the inputs change (although it already does that for mapState, it should just depend on mapState explicitly)

File: src/index.ts [hunk 2]
```diff
@@ -51,12 +48,6 @@ export function useMappedState<TState, TResult>(
     }
   };
 
-  if (prevStore !== store || prevMapState !== mapState) {
```
Comment (ianobermiller) [inline src/index.ts]: These can be removed in a separate PR if necessary.

File: src/index.ts [hunk 3]
```diff
@@ -23,6 +23,7 @@ const CONTEXT_ERROR_MESSAGE =
  */
 export function useMappedState<TState, TResult>(
   mapState: (state: TState) => TResult,
+  memoizationArray: Array<any> = [mapState],
```
Comment (ianobermiller) [inline src/index.ts]: Instead, default to null. Then, we can just wrap `mapState` in `useCallback`:

```tsx
const memoizedMapState = useCallback(mapState, memoizationArray);
```

**Comment by ianobermiller** [comment]
I think this would be a nice, non-breaking addition. We'll need to add some tests first, of course, and we'll probably want to update the documentation.

File: src/index.ts (hunk 1 above)
Comment (nmn) [reply src/index.ts]: So your idea is to wrap mapState in a useCallback and use that value for the useEffect?

Comment (ianobermiller) [reply src/index.ts]: Yup!

Comment (nmn) [reply src/index.ts]: I was thinking about maybe creating a memorized wrapper function within the hook such that you never need to resubscribe (unless the store changes) and this way it might be possible to ensure the top-down subscription order.

**Comment by ianobermiller** [comment]
See #9 and #25 for two other PRs that avoid resubscribing on mapState
change.

**Comment by ianobermiller** [comment]
@nmn Do you still want to get this PR cleaned up to merge?

**Comment by ianobermiller** [comment]
Alright, I know this is a popular request, but I consulted the React team and they recommended against custom hooks taking a dependency array. From @sebmarkbage:

> We discussed this last week and I think we settled on not recommending that custom hooks take deps.
>
> With a shallow compiler (as opposed to the Prepack strategy), we can only rewrite React components to more optimal form if they directly call a Hook in the react package with no indirection. So adding indirection would defeat those optimizations.
>
> It also adds a source of complexity to the whole ecosystem to have the linter be aware of every possible extension. Every app would have to come with a growing whitelist to enable the lint rule for those Hooks. It's also easy to miss since those that are excluded would just not warn. It really needs to be built-in to the type system at that point.
>
> It's also not very composable since every layer that adds more dependencies just grow.
>
> And even with the best form some of your inputs will be memoized in some cases anyway. Might be better to make that a pattern if needed.

I agree strongly with many of these points, particularly linting which I mentioned earlier. Good to hear also some notes about optimization and composability.
